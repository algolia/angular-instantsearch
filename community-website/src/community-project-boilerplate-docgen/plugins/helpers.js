const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
