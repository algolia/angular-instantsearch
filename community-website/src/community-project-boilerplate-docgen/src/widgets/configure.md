---
title: Widgets - Configure
layout: widget.pug
---

# Configure

## Description

Configure is a widget that lets you provide raw search parameters to the Algolia API.

Any of the searchParameters added to this widget will be forwarded to Algolia. For more information on the different parameters that can be set, have a look at the [reference](https://www.algolia.com/doc/api-client/javascript/search#search-parameters).

You can dynamically update the `searchParameters` Object and a new search will be triggered.

## Options

| Attribute            | Type       | Description
| -                    | -          | -
| `searchParameters`   | `Object`   | Search parameters to forward to Algolia

## Code example

```html
<ais-configure
  [searchParameters]="{ hitsPerPage: 3, analytics: true }"
>
</ais-configure>
```
