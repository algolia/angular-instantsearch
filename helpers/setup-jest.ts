import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
