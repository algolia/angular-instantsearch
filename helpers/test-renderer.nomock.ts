import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
