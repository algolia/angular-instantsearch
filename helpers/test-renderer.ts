import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
