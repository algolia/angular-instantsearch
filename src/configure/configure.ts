import {
  Component,
  Input,
  Inject,
  forwardRef,
  KeyValueDiffer,
  KeyValueDiffers,
} from '@angular/core';

import { connectConfigure } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import {
  NgAisInstantSearch,
  SearchParameters,
} from '../instantsearch/instantsearch';
import { noop } from '../utils';

export type ConfigureState = {
  refine: Function;
};

@Component({
  selector: 'ais-configure',
  template: '',
})
export class NgAisConfigure extends BaseWidget {
  // instance options
  private internalSearchParameters: SearchParameters;

  private differ: KeyValueDiffer<string, any>; // SearchParameters (I don't know how to get the values of the type)

  public state: ConfigureState = {
    refine: noop,
  };

  constructor(
    private differs: KeyValueDiffers,
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('Configure');
  }

  @Input()
  set searchParameters(values: SearchParameters) {
    this.internalSearchParameters = values;
    if (!this.differ && values) {
      this.differ = this.differs.find(values).create();
    }
  }

  public ngOnInit() {
    this.createWidget(connectConfigure, {
      searchParameters: this.internalSearchParameters,
    });
    super.ngOnInit();
  }

  ngDoCheck() {
    if (this.differ) {
      const changes = this.differ.diff(this.internalSearchParameters);
      if (changes) {
        this.state.refine(this.internalSearchParameters);
      }
    }
  }
}
