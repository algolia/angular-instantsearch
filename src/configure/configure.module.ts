import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisConfigure } from "./configure";

@NgModule({
  declarations: [NgAisConfigure],
  entryComponents: [NgAisConfigure],
  exports: [NgAisConfigure],
  imports: [CommonModule]
})
export class NgAisConfigureModule {}
