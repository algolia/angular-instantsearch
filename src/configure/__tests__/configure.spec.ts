import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisConfigure } from "../configure";

describe("Configure", () => {
  it("renders markup without state", () => {
    const render = createRenderer({
      TestedWidget: NgAisConfigure,
      template: "<ais-configure></ais-configure>"
    });
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });
});
