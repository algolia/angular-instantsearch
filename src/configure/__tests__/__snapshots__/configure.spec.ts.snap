// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configure renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgAisConfigure]}
>
  <ais-instantsearch>
    <ais-configure />
  </ais-instantsearch>
</ng-component>
`;
